Python and C++ code refactoring

carloslihu opened the pull request:
- Python code is now formatted with `black` and `isort`, and has been refactored according to `PEP 8` style guides.
- Python code partially commented with `google` docstring format.
- C++ code partially commented with `doxygen` docstring format.
- Scott's and Normal Reference Rule's `bandwidth` calculation have been reordered and commented.
- `ArcOperatorSet::update_incoming_arcs_scores` formulas have been reordered and commented.

davenza committed:
Release v0.5.1.

davenza committed:
Updates cibuildwheel to support manylinux.

carloslihu committed:
Python code refactored with "black"

carloslihu committed:
python imports sorted

carloslihu committed:
Merge remote-tracking branch 'upstream/master' into chore/refactoring

carloslihu committed:
extra files removed

carloslihu committed:
Pytests improved and partially commented

carloslihu committed:
util_test imports updated

carloslihu committed:
dataset/ commented

carloslihu committed:
util commented

carloslihu committed:
partial commenting

carloslihu committed:
more comments

carloslihu committed:
scotts and normal reference bandwidth recoded

carloslihu committed:
operators operations reordered

carloslihu committed:
pytest warning fixed

carloslihu committed:
extra comments cleaned

carloslihu committed:
0.5.2 changelog updated

carloslihu committed:
bandwidth covariances without the diagonal

carloslihu committed:
formatted_log_t verbose function

carloslihu committed:
is_psd with Cholesky decomposition

carloslihu committed:
diagonal covariance matrix commented

carloslihu committed:
try catch in bandwidth

carloslihu committed:
hill climbing verbose logs

carloslihu committed:
indentation fix

carloslihu committed:
set unix compilers

carloslihu committed:
ifdef warnings corrected

carloslihu committed:
typo

carloslihu committed:
DiscreteFactor_test fix

carloslihu committed:
code formatted

carloslihu committed:
pyproject.toml formatted

cli-titanium committed:
warning fix

cli-titanium committed:
eigen warnings fixed

cli-titanium committed:
util_test upgraded and renamed

cli-titanium committed:
Tests adapted: Variable naming to uppercase

cli-titanium committed:
util_test -> data

cli-titanium committed:
pytest -s

cli-titanium committed:
data.py reviewed

cli-titanium committed:
linear_correlation and mutual_info began

jcasajus-titanium committed:
independence_tests done

jcasajus-titanium committed:
chi square test

jcasajus-titanium committed:
independence tests documented

jcasajus-titanium committed:
Independence test summary

cli-titanium committed:
data MACROs

cli-titanium committed:
typo fix

cli-titanium committed:
warning ignored

JuanFPR-UPM committed:
Hybrid MI numerical limits fix

JuanFPR-UPM committed:
DiscreteFactor MLE Laplace Smoothing

carloslihu committed:
Merge pull request #1 from JuanFPR-UPM/Fix-Hybrid-MI-and-Discrete-MLE

Fix hybrid mi and discrete mle

JuanFPR-UPM committed:
MixedKCMI & VPTrees

JuanFPR-UPM committed:
Discrete Schurmann-Grassberger smoothing

cli-titanium committed:
Refactor PR with ms-vscode.cpptools

carloslihu committed:
Revert "Refactor PR with ms-vscode.cpptools"

This reverts commit 019cdbd4660b1950d35bc79f5e0ce1e5e697730e.

carloslihu committed:
Revert "Revert "Refactor PR with ms-vscode.cpptools""

This reverts commit a0d94d27b8d7f7707eebd12cd8c906cd77f35b83.

JuanFPR-UPM committed:
Added new files to cmake

JuanFPR-UPM committed:
Merge branch 'MixedKCMI-&-VPTrees' of https://github.com/JuanFPR-UPM/PyBNesian into MixedKCMI-&-VPTrees

carloslihu committed:
Merge pull request #2 from JuanFPR-UPM/MixedKCMI-&-VPTrees

Mixed kcmi & vp trees

carloslihu committed:
CMake Intellisense formatting

carloslihu committed:
Fix JSON formatting

JuanFPR-UPM committed:
Numerical fix MixedKCMI

carloslihu committed:
Merge pull request #3 from JuanFPR-UPM/MixedKCMI-&-VPTrees

Numerical fix MixedKCMI

carloslihu committed:
Add detailed installation guide for PyBNesian in INSTALLATION.md

carloslihu committed:
Merge branch 'master' of https://github.com/carloslihu/PyBNesian

carloslihu committed:
Merge remote-tracking branch 'origin/feature/ProductSPBN'

carloslihu committed:
Update pytest.ini to exclude 'vcpkg' directory from test discovery

carloslihu committed:
covariance matrix diagonal used for bandwidth selection

carloslihu committed:
Refactor test imports to use 'helpers' directory for data generation functions

carloslihu committed:
Update data generation functions: rename TRUE_LABEL and add generate_non_normal_data_classification

carloslihu committed:
Refactor tests with unified DATA_SIZE constant

carloslihu committed:
Refactor imports to use 'pbn' namespace for consistency across tests

carloslihu committed:
typo fixes

carloslihu committed:
Add function to generate discrete data classification with dependent variables

JuanFPR-UPM committed:
Hybrid MI numerical limits fix

carloslihu committed:
removed extra semicolon

carloslihu committed:
Merge pull request #4 from JuanFPR-UPM/Fix-Hybrid-MI-and-Discrete-MLE

Hybrid MI numerical limits fix

carloslihu committed:
Enable ccache for faster compilation if available

carloslihu committed:
Add Carlos Li Hu as a co-author in documentation and project metadata

carloslihu committed:
Rename TRUE_LABEL to TRUE_CLASS_LABEL for consistency in data generation functions

carloslihu committed:
fix: generate_discrete_data

carloslihu committed:
chore: update version to 0.5.2 in configuration files

carloslihu committed:
refactor: copilot documentation for KDE_test.py

carloslihu committed:
Merge remote-tracking branch 'origin/feature/diagonal-covariance'

carloslihu committed:
fix: update requirements to remove version constraints for pyarrow and pybind11

JuanFPR-UPM committed:
gamma approx fix

carloslihu committed:
Merge pull request #5 from JuanFPR-UPM:MixedKCMI-&-VPTrees

gamma approx fix

carloslihu committed:
docs: enhance installation instructions with verbose output and wheel creation steps

JuanFPR-UPM committed:
Fixed bug

carloslihu committed:
Merge pull request #6 from JuanFPR-UPM/MixedKCMI-&-VPTrees

Fixed bug

davenza commented:
Thank you for the PR @carloslihu . I am reviewing the PR. I don't understand the bug in hillclimbing.hpp (lines 152-160). Can you provide more detail?

I see some commented tests in tests/learning/algorithms/hillclimbing_test.py (I think they are related), but I do not know where is the "generate_normal_data_dep(1000)" (line 9) is defined.

carloslihu commented:
Hello David, thank you for reviewing my changes, they are a lot:

- The bug I'm talking about is about performing HC with cross-validation when you have a variable with low variability (but not 0 variance). For instance, a variable A with training values [0, 0, ..., 0, 1]. 
When performing the validated k-fold cross-validation, it finds problems fitting the distribution when the "1" value is not found in the training fold.
It is not a bug, more like a problem I found with some datasets. I do not have a clear solution for the algorithm. You may ignore the comment; I will remove it.

- I still have pending to review that the python tests are still working and cleaning comments. 
It's a minor change, but I have planned to finish this month.
Please wait for them, thanks

carloslihu commented:
Another issue is that I have retrocompatibility with the tests until the tag v0.5.2
However, afterwards I started using for KDE by default the diagonal covariance as the bandwidth matrix, instead of the full covariance (as suggested [here](https://bookdown.org/egarpor/NP-UC3M/kde-ii-mult.html))

This KDE learning faster and removes the Singular Covariance Matrix error.
Although it may return less accurate Bandwidth Matrices, in my experience, the differences are minimal.

If you like this change, I could adapt the tests so that they pass the results given this new method.

davenza commented:
Thank you @carloslihu. I will try to create a test with a case of almost 0 variance to try to control it. First, I wanna know where it is raising the error.

About the diagonal matrices: yes, many authors recommend using diagonal matrices. Less parameters to estimate and usually small degradation of performance. That is why I defined ProductKDE (https://pybnesian.readthedocs.io/en/latest/api/factors.html#pybnesian.ProductKDE). 

I know that ProductKDE is not exactly the same as KDE with diagonal bandwidth matrix, but: 

- Do you remember if the methods to estimate the bandwidth are equal (in ProductKDE and diagonal KDE)?
- If previous question is no. Do you think we need an specific implementation for the estimation of bandwidth in diagonal KDEs?

carloslihu commented:
Hello David,
After analyzing the definition of [PyBNesian ProductKDE](https://pybnesian.readthedocs.io/en/latest/api/factors.html#pybnesian.ProductKDE) and the definition of [diagonal bandwidth matrix KDE](https://bookdown.org/egarpor/NP-UC3M/kde-ii-mult.html) found in the literature.
I think they represent the same concept with different notation.

For my implementation I wanted to use ProductKDE for Hill Climbing, but by lack of skill and time, I wasn't able to do so.
Therefore, my **Hack Fix** was to manually remove the non-diagonal elements from the covariance matrix in
[NormalReferenceRule.hpp](https://github.com/carloslihu/PyBNesian/blob/320ad4a179a7051ea001b338a77b4b11127c6d3a/pybnesian/kde/NormalReferenceRule.hpp#L168-L175) and [ScottsBandwidth.hpp](https://github.com/carloslihu/PyBNesian/blob/320ad4a179a7051ea001b338a77b4b11127c6d3a/pybnesian/kde/ScottsBandwidth.hpp#L142-L149)

This is just a temporary fix, I think that the best solution would be to allow the option of allowing the selection of your ProductKDE implementation for the Hill Climbing algorithm, but my lack of knowledge wouldn't let me.

carloslihu commented:
Hello David, 
For the moment I have rebased my master branch to tag v0.5.2 (before adapting my diagonal bandwidth fix)
This way you may merge my latest changes without altering your functionality.

I will continue using the diagonal bandwidth in another personal branch (feature/diagonal-bandwidth) for the moment

carloslihu committed:
fix: clean up formatting and improve consistency in documentation files

JuanFPR-UPM committed:
Parallelized implementation

carloslihu committed:
Merge pull request #7 from JuanFPR-UPM/MixedKCMI-&-VPTrees

Mixed kcmi & vp trees

carloslihu committed:
Merge remote-tracking branch 'origin/master' into feature/diagonal-bandwidth

carloslihu committed:
fix: update README with diagonal-bandwidth instructions

carloslihu committed:
Merge branch 'master' into feature/diagonal-bandwidth

carloslihu committed:
chore: update version to 0.5.3 and improve documentation

carloslihu committed:
feat: add training type support to LinearGaussianCPD and update tests for float32 sampling

carloslihu committed:
docs: update installation guide with improved instructions and additional dependencies

carloslihu committed:
feat: add delta index comparator to fix seed in hc learning

carloslihu committed:
Refactor variable names in serialization tests to lowercase

- Updated all instances of uppercase variable names (A, B, C, D) to lowercase (a, b, c, d) in the serialization tests for various Bayesian network models.
- Ensured consistency across fixtures and assertions to reflect the new naming convention.

carloslihu committed:
feat: refactor KDE and CKDE tests to use diagonal bandwidth calculations

carloslihu committed:
chore: update version to 0.6.0 and enhance changelog

carloslihu committed:
fix: update arrow dependency format to include default features

carloslihu committed:
feat: add utf8proc port with CMake configuration and vcpkg integration

carloslihu committed:
fix: update CMakeLists.txt for proper include and install paths

carloslihu committed:
chore: remove utf8proc port files as part of cleanup

carloslihu committed:
feat: add vcpkg logs upload step on failure in release workflows

carloslihu committed:
feat: add vcpkg logs handling in build workflows

carloslihu committed:
fix: correct command syntax for git submodule update in CMakeLists.txt
